wagtail/core/query.py

```diff
@@ -237,6 +272,71 @@ def not_public(self):
         """
         return self.exclude(self.public_q())
 
+    def _get_closest_view_restriction(self, page):
+        """
+        Return the PageViewRestriction most applicable to the provided ``page`` instance.
+
+        View restrictions apply to the selected page AND their descendants, so for each
+        page, it's the restriction applied to the 'closest' ancestor that determines
+        what restrictions apply to it.
+        """
+        for vr in self.all_view_restrictions:
+            if page.id == vr.page.id or page.is_descendant_of(vr.page):
+                return vr
+
+    def for_user(self, user=None, include_password_protected=False):
+        """
+        Return a queryset of public and private pages that should be
+        accessible to the provided ``user``, according to their group
+        memberships and current page view restriction settings.
+
+        It is impossible to know whether the user knows the password
+        to access a password-protected page/section, so those pages are
+        excluded by default. However, you can use the
+        ``include_password_protected`` keyword argument to override this
+        behaviour and include them if needed.
+        """
```

Comment:
I wonder if we could cache the list of protected paths that apply to this user on the request?

Eg, the first time this is called, it fetches all page view restrictions and removes any that user passes. Then saves the page path's that are still restricted for subsequent calls.

Is there any particular reason why we're doing this for just the user and not the whole request? (that would also help us with the password-restricted views)

Reply:
In the project I implemented this on, it's only used in a couple of places (once per request), so there's no real advantage to bringing requests into it. However, I can totally see the benefit of that in a Wagtail-native solution. This has all of a sudden become much more exciting :D thanks @kaedroho